Sparse index: integrate with `read-tree`

vdye opened the pull request:
This pull request is more-or-less split into "fixups" and "integrating sparse index with `read-tree`".

## Fixups
- Fix syntax in some of the `diff` tests (I can't seem to get `echo` to work in the tests, so it's replaced with a custom `edit-contents`)
- Fix `git status` reporting differences nested in sparse directory
  - In the test relevant to this commit, `git status` was previously reporting changes nested inside of a sparse directory as the whole directory being "modified" (but _only_ in the sparse index). Fixed by setting the `recursive` flag when performing the diff
- Modify the way `prime_cache_tree_rec` identifies sparse directories when rebuilding the cache tree
  - The previous implementation was breaking when `read-tree` was sparse-enabled in the `t1091` tests. The prior checks used weren't correctly identifying sparse cache entries in the case that the sparse patterns were broken.
- Change the test looking for sparse index expand and collapse to use `git mv` (which, as far as I can tell, isn't slated for integration with sparse index in the immediate future)
  - This one is a `fixup???` since the commit message for the original would need to be changed, since it calls out `read-tree` by name

## `read-tree`
Other than the standard changes to enable sparse index:
- Expand the index when `--no-sparse-checkout` is provided
  - The option effectively removes the `skip-worktree` flag from all entries, which eventually leads to the command attempting to checkout a sparse directory
- Expand the index when the prefix is equivalent to or is nested inside a sparse directory
  -  When `--prefix` is specified, it forces the `traverse_path` in `unpack_trees`/`traverse_trees` to start at the given prefix. However, the operation of those functions implicitly assumes (requires?) that the traverse path is something it has actually traversed into (which it can't be for a sparse directory). So, if the prefix is a sparse directory (e.g., `folder1` in the `t1092` tests) or is nested in one (e.g., `folder2/0`), the prefix is expanded before continuing with `unpack_trees`

## Performance testing
```
Test                                                   ms/vfs-2.33.0     HEAD                  
-----------------------------------------------------------------------------------------------
2000.2: git status (full-v3)                           0.74(0.28+0.37)   0.97(0.35+0.44) +31.1%
2000.3: git status (full-v4)                           0.80(0.30+0.39)   0.92(0.46+0.40) +15.0%
2000.4: git status (sparse-v3)                         0.56(0.11+0.69)   0.50(0.10+0.71) -10.7%
2000.5: git status (sparse-v4)                         0.54(0.12+0.80)   0.51(0.11+0.72) -5.6% 
2000.22: git reset (full-v3)                           0.78(0.35+0.36)   0.77(0.35+0.27) -1.3% 
2000.23: git reset (full-v4)                           0.84(0.47+0.25)   0.90(0.49+0.29) +7.1% 
2000.24: git reset (sparse-v3)                         0.36(0.07+0.77)   0.46(0.09+1.09) +27.8%
2000.25: git reset (sparse-v4)                         0.40(0.09+0.65)   0.44(0.09+0.63) +10.0%
2000.26: git reset --hard (full-v3)                    1.03(0.61+0.39)   1.11(0.62+0.43) +7.8% 
2000.27: git reset --hard (full-v4)                    1.12(0.65+0.40)   1.18(0.70+0.41) +5.4% 
2000.28: git reset --hard (sparse-v3)                  0.51(0.08+0.31)   0.63(0.10+0.39) +23.5%
2000.29: git reset --hard (sparse-v4)                  0.49(0.08+0.30)   0.56(0.09+0.33) +14.3%
2000.30: git read-tree -mu HEAD (full-v3)              1.29(0.96+0.30)   1.45(1.12+0.34) +12.4%
2000.31: git read-tree -mu HEAD (full-v4)              1.28(0.96+0.30)   1.41(1.10+0.33) +10.2%
2000.32: git read-tree -mu HEAD (sparse-v3)            1.66(1.25+0.40)   0.13(0.07+0.05) -92.2%
2000.33: git read-tree -mu HEAD (sparse-v4)            1.61(1.22+0.38)   0.13(0.07+0.05) -91.9%
```

vdye committed:
fixup! sparse-index: use `read-tree` for expand/collapse test

Change the command used in the test `sparse-index is expanded and converted
back` to `git mv` so the test continues to pass after integrating `read-tree`
with sparse index.

Signed-off-by: Victoria Dye <vdye@github.com>

vdye committed:
read-tree: expand sparse checkout test coverage

Add tests primarily focused on `read-tree` merges outside the sparse checkout
definition. One test appears to indicate a pre-existing bug, and is included
for documentation as a `test_expect_failure` test.

Signed-off-by: Victoria Dye <vdye@github.com>

vdye committed:
status: bugfix directory diff in sparse index

Adding the `recursive` flag for `wt_status_collect_changes_index` allows the
diff function to recurse into sparse directories when differences occur within
them. This corrects the issue observed in the `read-tree` sparse compatibility
test in which particular merges resulted in a directory appearing as "modified"
for sparse index repositories, rather than showing only the modified file
within the directory.

The `recursive` flag, even with a non-sparse index, does not substantially
change the performance of `status`. From the `p2000` performance suite:

Test                             before            after
------------------------------------------------------------------------
2000.2: git status (full-v3)     0.68(0.27+0.34)   0.64(0.26+0.33) -5.9%
2000.3: git status (full-v4)     0.68(0.27+0.35)   0.67(0.26+0.34) -1.5%
2000.4: git status (sparse-v3)   0.45(0.09+0.63)   0.45(0.10+0.74) +0.0%
2000.5: git status (sparse-v4)   0.45(0.09+0.62)   0.45(0.10+0.73) +0.0%

Signed-off-by: Victoria Dye <vdye@github.com>

vdye committed:
fixup! diff: enable and test the sparse index

Update tests for diff to successfully `echo`.

Signed-off-by: Victoria Dye <vdye@github.com>

vdye committed:
read-tree: integrate with sparse index

Add repository setting setup for enabling sparse index usage with `read-tree`.
In order to maintain behavior (and allow for focused adjustment of
implementation in later commits), index is expanded when `--no-sparse-checkout`
or `--prefix` is specified.

Substantial performance improvement for sparse index indicated by `p2000`
results:

Test                                 before            after
-----------------------------------------------------------------------------
git read-tree -mu HEAD (full-v3)     1.21(0.94+0.30)   1.37(1.06+0.33) +13.2%
git read-tree -mu HEAD (full-v4)     1.18(0.90+0.28)   1.35(1.06+0.31) +14.4%
git read-tree -mu HEAD (sparse-v3)   1.61(1.23+0.40)   0.14(0.07+0.05) -91.3%
git read-tree -mu HEAD (sparse-v4)   2.16(1.52+0.48)   0.14(0.07+0.05) -93.5%

Signed-off-by: Victoria Dye <vdye@github.com>

vdye committed:
read-tree: narrow scope of index expansion for `--prefix`

When `unpack_trees` is provided with a prefix, the tree traversal that occurs
in it requires that the base directory path given to it is a traversable
directory in the index. In order to adhere to this requirement, the index is
expanded if the given prefix is equal to or a subdirectory of a sparse
directory in the index.

Signed-off-by: Victoria Dye <vdye@github.com>

vdye committed:
fixup! sparse-index: make non-mixed `reset` sparse-aware

Fix an edge case where the cache tree is constructed improperly because invalid
cone-mode patterns are manually set in the sparse definition. Also fix test bug
(would not report error without `|| return 1`).

Signed-off-by: Victoria Dye <vdye@github.com>

vdye committed:
read-tree: add performance test in p2000

Signed-off-by: Victoria Dye <vdye@github.com>

derrickstolee commented:
> ## Fixups
> * Fix syntax in some of the `diff` tests (I can't seem to get `echo` to work in the tests, so it's replaced with a custom `edit-contents`)

The redirection character doesn't work as it encompasses the entire `test_all_match` function call. Unfortunately the `edit-contents` script is our best approach right now.

> * Fix `git status` reporting differences nested in sparse directory
>   
>   * In the test relevant to this commit, `git status` was previously reporting changes nested inside of a sparse directory as the whole directory being "modified" (but _only_ in the sparse index). Fixed by setting the `recursive` flag when performing the diff

Wow! Good catch.

> * Modify the way `prime_cache_tree_rec` identifies sparse directories when rebuilding the cache tree
>   
>   * The previous implementation was breaking when `read-tree` was sparse-enabled in the `t1091` tests. The prior checks used weren't correctly identifying sparse cache entries in the case that the sparse patterns were broken.

I'll go read the commit for this, but it sounds like we can get into strange situations where the sparse-checkout patterns change to no longer be cone-mode, but we still have a sparse index. That's a good case to handle.

> * Change the test looking for sparse index expand and collapse to use `git mv` (which, as far as I can tell, isn't slated for integration with sparse index in the immediate future)

Good idea. It's not high on the list. We might eventually drop the "definitely expand" test, but for now we can continue chasing commands that still expand it.

>   * This one is a `fixup???` since the commit message for the original would need to be changed, since it calls out `read-tree` by name

Right. We could squash it in when upstreaming that commit. You could also do a `git commit --fixup <oid>` followed by a `git commit --fixup=reword:<oid>` to modify the commit message and we will naturally squash them together with the new message in the 2.34 rebase.

> ## `read-tree`
> Other than the standard changes to enable sparse index:
> 
> * Expand the index when `--no-sparse-checkout` is provided
>   
>   * The option effectively removes the `skip-worktree` flag from all entries, which eventually leads to the command attempting to checkout a sparse directory

Makes sense.

> * Expand the index when the prefix is equivalent to or is nested inside a sparse directory
>   
>   * When `--prefix` is specified, it forces the `traverse_path` in `unpack_trees`/`traverse_trees` to start at the given prefix. However, the operation of those functions implicitly assumes (requires?) that the traverse path is something it has actually traversed into (which it can't be for a sparse directory). So, if the prefix is a sparse directory (e.g., `folder1` in the `t1092` tests) or is nested in one (e.g., `folder2/0`), the prefix is expanded before continuing with `unpack_trees`

This also makes sense for safety. We could extend this later to refuse to call `unpack_trees()` if the prefix is within a sparse directory (or generally outside of the sparse-checkout definition). But, I doubt that will actually be worth the effort since users shouldn't be specifying the prefix this way.

> ## Performance testing
> ```
> Test                                                   ms/vfs-2.33.0     HEAD                  
> -----------------------------------------------------------------------------------------------
> 2000.30: git read-tree -mu HEAD (full-v3)              1.29(0.96+0.30)   1.45(1.12+0.34) +12.4%
> 2000.31: git read-tree -mu HEAD (full-v4)              1.28(0.96+0.30)   1.41(1.10+0.33) +10.2%
> 2000.32: git read-tree -mu HEAD (sparse-v3)            1.66(1.25+0.40)   0.13(0.07+0.05) -92.2%
> 2000.33: git read-tree -mu HEAD (sparse-v4)            1.61(1.22+0.38)   0.13(0.07+0.05) -91.9%
> ```

Feel free to delete the unnecessary rows (like this) when talking about them in your commit messages.

Now... to go looking at the code.

derrickstolee commented on read-cache.c:
```diff
@@ -611,6 +611,28 @@ int index_name_pos(struct index_state *istate, const char *name, int namelen)
 	return index_name_stage_pos(istate, name, namelen, 0);
 }
 
+int index_entry_exists(struct index_state *istate, const char *name, int namelen)
```
This implementation _could_ be `return index_name_stage_pos() >= 0;` but we specifically don't want to expand the index in the case of a sparse directory. Perhaps it would be better to extend `index_name_stage_pos()` to have an extra parameter that controls that expansion. Then `index_entry_exists()` could still use that same binary search.

It would be worth commenting on the declaration of `index_entry_exists()` that it does not expand a sparse index and hence will return 0 for paths within sparse directory entries.

vdye replied:
Thanks, that certainly helped clean up some code duplication!

derrickstolee commented on wt-status.c:
```diff
@@ -651,6 +651,14 @@ static void wt_status_collect_changes_index(struct wt_status *s)
 	rev.diffopt.rename_limit = s->rename_limit >= 0 ? s->rename_limit : rev.diffopt.rename_limit;
 	rev.diffopt.rename_score = s->rename_score >= 0 ? s->rename_score : rev.diffopt.rename_score;
+
+	/*
+	 * The `recursive` flag must be set to properly perform a diff on sparse
+	 * directory entries.
+	 */
+	if (s->repo->index->sparse_index)
+		rev.diffopt.flags.recursive = 1;
```
I wonder if there is any negative to setting `recursive = 1` in all cases. My guess is that it will be correct, but possibly slower? It would be great if it wasn't actually any slower and we don't need the special case.

vdye replied:
I'll check this (and add the performance numbers to the commit message) once the functional issues in this series are ironed out. For now, I'll leave this conversation open so I don't forget about it. 🙂

vdye replied:
It turns out that, even on full-index `git status` the `recursive` flag didn't affect performance in any noticeable way! Probably because a `git status` diff isn't normally running a diff on a directory anyway. Sparse index check removed + performance results added in commit message.

derrickstolee commented on t/t1092-sparse-checkout-compatibility.sh:
```diff
@@ -846,12 +846,7 @@ test_expect_success 'read-tree --prefix outside sparse definition' '
-# as modified in the sparse index, but `folder2/0/a` (correctly?) appears
-# instead as a modified entry in the non-sparse index checkouts.
-test_expect_failure 'read-tree --merge with directory-file conflicts' '
+test_expect_success 'read-tree --merge with directory-file conflicts' '
```
Great job finding and fixing this bug!

derrickstolee commented on builtin/read-tree.c:
```diff
@@ -214,6 +217,10 @@ int cmd_read_tree(int argc, const char **argv, const char *cmd_prefix)
 
+	/* TODO: audit sparse index behavior in unpack_trees */
+	if (opts.skip_sparse_checkout || opts.prefix)
+		ensure_full_index(&the_index);
```
A very good approach so far. These are reasonable protections.

derrickstolee commented on t/t1092-sparse-checkout-compatibility.sh:
```diff
@@ -1230,6 +1230,19 @@ test_expect_success 'sparse index is not expanded: update-index' '
+test_expect_success 'sparse index is not expanded: read-tree' '
+	init_repos &&
+
+	ensure_not_expanded checkout -b test-branch update-folder1 &&
+	for MERGE_TREES in "base HEAD update-folder2" \
+		      "HEAD update-folder2" \
+		      "update-folder2"
+	do
```
One case that might be interesting is a situation where we do a two- or three-way merge and two of the refs changed files such as `folder1/a` and `folder1/b`, respectively. This would make them look like they have an edit-edit conflict on the sparse directory entry, but if we expand the directory there would not be such a conflict.

vdye replied:
Ah, I completely missed this case (and you're right - it does cause a conflict). Since the conflict is only detected once you're already in `twoway_merge` or `threeway_merge` (in the middle of `traverse_trees`), you wouldn't be able to expand the index to resolve that conflict. 

I _think_ one way to handle this might be the way `oneway_diff` does (using `diff_tree_oid` to get the differences at a file level, then reconstructing the two- and three-way merges from series of one-way diffs). Since you wrote that addition to `oneway_diff`, I wanted to check - does that make sense? Or were you thinking something different for solving this?

derrickstolee replied:
I _think_ that's the best way to attempt handling this. The tricky part is constructing the trees again afterwards, but maybe you can plug into the merge machinery to merge the trees to get a new tree? The ORT API might be a way to go here.

The other idea would be to report that the operation failed due to a sparse directory, then expand and try again. This isn't a great idea, but would be simpler to implement.

vdye replied:
I added a _tentative_ implementation for handling two-way merges (which, ultimately, are merging three trees: the index, input_tree_a, and input_tree_b). It passes the tests, but I'm still worried about all of the different edge cases of merges that could happen (e.g., `merge-ort` might successfully merge the contents of a file modified in two branches, but the existing `twoway_merge` might not). And, I haven't added anything yet for `threeway_merge` since the `n` number of base trees (very specific index conditions) don't currently fit nicely with `merge-ort`.

I do still want to avoid excessive index expansion, though - I _think_ one possibility is to refactor `unpack_trees` itself to, rather than using an element-by-element merge function, use `merge-ort` directly (when `unpack_trees_options.merge` is set). It would also require an extension to `merge-ort` to handle multiple base trees in a non-recursive merge. If it's not a terrible idea, I think it's doable - however, it's pretty substantial and probably warrants its own branch/patch series.

In any case, please look over what I have and let me know if you have any suggestions on direction. It's also potentially possible to wait on the two-way and three-way merge cases, since it looks like stash only ever calls `read-tree` with one tree argument.

derrickstolee commented on t/t1092-sparse-checkout-compatibility.sh:
```diff
@@ -1240,7 +1240,16 @@ test_expect_success 'sparse index is not expanded: read-tree' '
-	done
+	done &&
+
+	ensure_not_expanded checkout -b test-branch-2 &&
+	mkdir -p sparse-index/deep/test-dir &&
+	cp sparse-index/a sparse-index/deep/test-dir &&
+	ensure_not_expanded add deep/test-dir &&
+	ensure_not_expanded commit -m "test" &&
+
+	ensure_not_expanded checkout test-branch &&
+	ensure_not_expanded read-tree --prefix=deep/test-dir test-branch-2
```
Do you want to try something with `--prefix=deep/deeper2` instead of creating this `test-dir`?

vdye replied:
When `read-tree`-ing with only one tree argument, `bind_merge(...)` prevents "merging" when a file already exists within the prefix. If I run `git read-tree --prefix=deep/deeper2 update-deep`, I get the error:

```
error: Entry 'deep/deeper2/a' overlaps with 'deep/deeper2/a'.  Cannot bind.
```

To avoid creating the new directory, though, I adjusted the test to instead remove the files in `deep/deeper2`, commit that to the test branch, then run `git read-tree --prefix=deep/deeper2 deepest`, which ends up testing the same(ish) scenario.

derrickstolee commented on unpack-trees.c:
```diff
@@ -1741,6 +1741,51 @@ int unpack_trees(unsigned len, struct tree_desc *t, struct unpack_trees_options
 	}
 
+	/*
+	 * If the prefix is or is contained within a sparse directory, the index
```
"If the prefix is equal to or contained within..." 

My first read of "is or is" was jarring.

derrickstolee commented on unpack-trees.c:
```diff
@@ -1741,6 +1741,51 @@ int unpack_trees(unsigned len, struct tree_desc *t, struct unpack_trees_options
+				if (!S_ISSPARSEDIR(ce->ce_mode))
+					continue;
+
+				/*
+				 * Normalize comparison length for cache entry vs. prefix -
+				 * either may have a trailing slash, which we do not want to
+				 * compare (can assume both are directories).
+				 */
+				if (ce->name[ce_len - 1] == '/')
+					ce_len--;
+				if (o->prefix[prefix_len - 1] == '/')
+					prefix_len--;
```
Oddly, we _want_ the paths to end with a slash, because "A/Dir" is a prefix of "A/Directory" but "A/Dir/" is not a prefix of "A/Directory/". One way to check for this is actually to check that `ce_len == prefix_len || o->prefix[ce_len] == '/'` in the same condition as the `!strcmp()`.

vdye replied:
I went with an `if (opposite of that condition) continue;` block (to mirror the style of the rest of the block, but otherwise this should be fixed. Good catch!

derrickstolee replied:
DeMorgan to the rescue!

derrickstolee commented on t/perf/p2000-sparse-operations.sh:
```diff
@@ -112,6 +112,7 @@ test_perf_on_all git commit -a -m A
 test_perf_on_all git checkout -f -
 test_perf_on_all git reset
 test_perf_on_all git reset --hard
+test_perf_on_all git read-tree -mu HEAD
```
If you reorder this to be early in the series, you can then advertise the results of this test in the commit message for the commit that improves the performance.

vdye replied:
I moved the commit, but haven't added performance numbers yet (like with the `git status` change, I'll add those once any outstanding bugs are fixed). In the meantime, I'll keep this open to remind myself.

vdye replied:
Performance numbers added to commit message - ~90% reduction in execution time!

derrickstolee reviewed (commented):
Great start. There are some interesting bits in here, and I really appreciate the fixes for previously-merged code.

I have some recommendations to look into new test cases or tricky situations.

vdye committed:
amend! sparse-index: use `read-tree` for expand/collapse test

sparse-index: update command for expand/collapse test

In anticipation of multiple commands being fully integrated with sparse index,
update the test for index expand and collapse for non-sparse index integrated
commands to use `mv`.

Signed-off-by: Victoria Dye <vdye@github.com>

derrickstolee commented on unpack-trees.c:
```diff
@@ -2512,6 +2519,55 @@ static void show_stage_entry(FILE *o,
+		merge = dup_cache_entry(left, o->src_index);
+		oidcpy(&merge->oid, &result.tree->object.oid);
+		ret = merged_entry(merge, left, o);
+	}
+	else {
```
nit: `} else {`

derrickstolee commented on unpack-trees.c:
```diff
@@ -2512,6 +2519,55 @@ static void show_stage_entry(FILE *o,
+		ret = merged_entry(merge, left, o);
+	}
+	else {
+		/*
+		 * Find the first conflicted entry, reject the merge for that.
+		 */
```
Nit: single-line comments will want to be on a single line.

```
/* Find the first conflicted entry, reject the merge for that. */
```

derrickstolee commented on merge-ort.c:
```diff
@@ -4527,6 +4527,15 @@ static void merge_ort_nonrecursive_internal(struct merge_options *opt,
 	result->tree = parse_tree_indirect(&working_tree_oid);
 	/* existence of conflicted entries implies unclean */
 	result->clean &= strmap_empty(&opt->priv->conflicted);
+	if (!strmap_empty(&opt->priv->conflicted))
+	{
```
nit: `if (...) {`

It's a hard habit to re-learn.

derrickstolee commented on builtin/read-tree.c:
```diff
@@ -223,11 +230,21 @@ int cmd_read_tree(int argc, const char **argv, const char *cmd_prefix)
 			break;
 		case 3:
 		default:
+			/*
+			 * TODO: update threeway_merge to handle edit/edit conflicts in
+			 * sparse directories.
+			 */
+			ensure_full_index(&the_index);
```
My guess is that this case is rare enough that we might not even bother with it. Let's see if `git stash` ever needs this case, for instance. This `ensure_full_index()` might be a long-term solution.

vdye replied:
Since `git stash` only needs `oneway_diff` (single-tree) cases to work, I'll leave the `ensure_full_index()` calls in for two-way and three-way merges.

derrickstolee commented on t/t1092-sparse-checkout-compatibility.sh:
```diff
@@ -1250,6 +1250,17 @@ test_expect_success 'sparse index is not expanded: update-index' '
+	init_repos &&
+
+	ensure_not_expanded checkout -b test-branch update-folder1 &&
+	for MERGE_TREES in "update-folder2"
```
This is a nice way to prepare for showing new cases in the future.

derrickstolee commented on t/t1092-sparse-checkout-compatibility.sh:
```diff
@@ -1254,7 +1254,8 @@ test_expect_success 'sparse index is not expanded: read-tree' '
 
 	ensure_not_expanded checkout -b test-branch update-folder1 &&
-	for MERGE_TREES in "update-folder2"
+	for MERGE_TREES in "update-folder1 update-folder2" \
+		      "update-folder2"
```
nit: this whitespace doesn't look right (on GitHub, maybe it's right in the editor).

```
	for MERGE_TREES in "update-folder1 update-folder2" \
			   "update-folder2"
```

derrickstolee reviewed (approved):
I like your new approach. I only found style-based nitpicks.

I think leaving the last `ensure_full_index()` is something we can handle for now, unless we find it is critical to the success of the `git stash` work.

vdye commented on unpack-trees.c:
```diff
@@ -2512,6 +2519,55 @@ static void show_stage_entry(FILE *o,
+	left_tree = lookup_tree(o->src_index->repo, &left->oid);
+	right_tree = lookup_tree(o->src_index->repo, right ? &right->oid : empty_tree_oid);
+
+	merge_incore_nonrecursive(&opt, base_tree, left_tree, right_tree, &result);
```
@derrickstolee After thinking about this more, `merge-ort` isn't _quite_ right for this case (as it is now, anyway) - `merge-ort` performs content merges, whereas `twoway_merge` does not and has its own set of rules for how to "merge" files. Since `stash` only uses the `oneway_merge` case of `read-tree`, I'd like to revert back to d25d0f2e1020e97eea8c51acfef86b9286392113 (with the code style cleanup you've pointed out). That should still get the performance benefit in `stash` (I'll verify that with the tests, though), as well as not introducing functionality that creates mismatches between sparse index and non-sparse index checkouts.

derrickstolee replied:
That works for me. Let's not make any unnecessary changes here.
